File: src/main/java/com/example/gpiApp/service/DashboardLayoutService.java
package com.example.gpiApp.service;

import com.example.gpiApp.entity.DashboardLayout;
import com.example.gpiApp.repository.DashboardLayoutRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

/**
 * Reads and upserts a user's dashboard layout (opaque JSON of widget keys). One row per user; created
 * on first save. The widget vocabulary lives in the frontend — the backend just stores the string.
 */
@Service
@RequiredArgsConstructor
public class DashboardLayoutService {

    private static final int MAX_LEN = 4000;

    private final DashboardLayoutRepository repository;

    @Transactional(readOnly = true)
    public String getWidgets(Long userId) {
        return repository.findByUserId(userId).map(DashboardLayout::getWidgets).orElse(null);
    }

    @Transactional
    public String save(Long userId, String widgets) {
        if (widgets != null && widgets.length() > MAX_LEN)
            throw new IllegalArgumentException("Layout is too large.");
        DashboardLayout layout = repository.findByUserId(userId)
                .orElseGet(() -> DashboardLayout.builder().userId(userId).build());
        layout.setWidgets(widgets);
        return repository.save(layout).getWidgets();
    }
}

File: frontend/src/app/core/services/dashboard-layout.service.ts
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { ApiService } from './api.service';

export interface DashboardLayout {
  widgets: string[];
}

/** Per-user customizable dashboard layout (ordered KPI widget keys). */
@Injectable({ providedIn: 'root' })
export class DashboardLayoutService {
  private base = '/dashboard-layout';
  constructor(private api: ApiService) {}

  get(): Observable<DashboardLayout> { return this.api.get<DashboardLayout>(this.base); }
  save(widgets: string[]): Observable<DashboardLayout> { return this.api.put<DashboardLayout>(this.base, { widgets }); }
}
